project/src/Agents/GSTNode2DAgent.gd

```diff
@@ -0,0 +1,165 @@
+# tags: virtual
+func _apply_steering(acceleration: GSTTargetAcceleration, delta: float) -> void:
+	_applied_steering = true
+	match _body_type:
+		BodyType.RIGID:
+			_apply_rigid_steering(acceleration)
+		BodyType.KINEMATIC:
+			match kinematic_movement_type:
+				MovementType.COLLIDE:
+					_apply_collide_steering(acceleration.linear, delta)
+				MovementType.SLIDE:
+					_apply_sliding_steering(acceleration.linear)
+				MovementType.POSITION:
+					_apply_normal_steering(acceleration.linear, delta)
+		BodyType.NODE:
+			_apply_normal_steering(acceleration.linear, delta)
+	if not _body_type == BodyType.RIGID:
+		_apply_orientation_steering(acceleration.angular, delta)
+
+
+func _apply_rigid_steering(accel: GSTTargetAcceleration) -> void:
+	body.apply_central_impulse(GSTUtils.to_vector2(accel.linear))
+	body.apply_torque_impulse(accel.angular)
+	if calculate_velocities:
+		linear_velocity = GSTUtils.to_vector3(body.linear_velocity)
+		angular_velocity = body.angular_velocity
+
+
+func _apply_sliding_steering(accel: Vector3) -> void:
+	var velocity := GSTUtils.to_vector2(linear_velocity + accel).clamped(linear_speed_max)
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(Vector2.ZERO, linear_drag_percentage)
+	velocity = body.move_and_slide(velocity)
+	if calculate_velocities:
+		linear_velocity = GSTUtils.to_vector3(velocity)
+
+
+func _apply_collide_steering(accel: Vector3, delta: float) -> void:
+	var velocity := GSTUtils.clampedv3(linear_velocity + accel, linear_speed_max)
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(
+				Vector3.ZERO,
+				linear_drag_percentage
+		)
+	body.move_and_collide(GSTUtils.to_vector2(velocity) * delta)
+	if calculate_velocities:
+		linear_velocity = velocity
+
+
+func _apply_normal_steering(accel: Vector3, delta: float) -> void:
+	var velocity := GSTUtils.clampedv3(linear_velocity + accel, linear_speed_max)
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(
+				Vector3.ZERO,
+				linear_drag_percentage
+		)
+	body.global_position += GSTUtils.to_vector2(velocity) * delta
+	if calculate_velocities:
+		linear_velocity = velocity
+
+
+func _apply_orientation_steering(angular_acceleration: float, delta: float) -> void:
+	var velocity = angular_velocity + angular_acceleration
+	if apply_angular_drag:
+		velocity = lerp(velocity, 0, angular_drag_percentage)
+	body.rotation += velocity * delta
+	if calculate_velocities:
+		angular_velocity = velocity
+
+
+func _set_use_physics(value: bool) -> void:
+	if use_physics and not value:
+		body.get_tree().disconnect("idle_frame", self, "_on_SceneTree_frame")
+		body.get_tree().connect("physics_frame", self, "_on_SceneTree_frame")
+	elif not use_physics and value:
+		body.get_tree().disconnect("physics_frame", self, "_on_SceneTree_frame")
+		body.get_tree().connect("idle_frame", self, "_on_SceneTree_frame")
+	use_physics = value
+
+
+func _set_body(value: Node2D) -> void:
+	body = value
+	if body is RigidBody2D:
```

Comment:
Is type check costly? Because if not, I'd remove the extra _body_type variable and use type checks throughout the class, removing the need for some of the code.

Reply:
Having just run a test:
```gdscript
for i in range(1000000):
    if self is Node:
        increment += 1

var test = if self is Node
for i in range(1000000):
    if test:
        increment += 1
```

The former takes 0.000169 seconds,
the latter takes 0.000068

So on the time scale we're talking about, it seems to be fairly negligible until you have a lot of those agents all doing those tests.

As for using steering behaviors with steering agents, absolutely. We used them back in Voidspace, even though we had Box2D rigid bodies. Now, whether someone will use the specialized agent and use apply_steering rather than roll their own solution, that's up in the air, but I don't see why not.

Reply:
Okay all good, you can ignore other comments that bring up the same point then!